### PR TITLE
feat: Add dist support for packaging

### PR DESCRIPTION
**Pull request by imjoey**
Related to #64. 

This PR is going to add support for generating the `dist` for both rpm and deb artifacts. It's well noted that the difference of artifacts is only determined by the building phase in different OS distributions. The implementation is to automatically determine the `dist` according to the container OS distribution (by `image_base` and `image_tag`) during the building phase. Please see https://github.com/api7/apisix-build-tools/pull/67#issuecomment-896536098 for details.

Once this PR is merged, the CI scripts in several repositories should be also updated accordingly. And then we could add more CI cases to generate `rpm`/`deb` to support many other OS distributions.

Signed-off-by: imjoey <majunjie@apache.org>

### CODE REVIEW

File: Makefile
Comment (EnableAsync) [inline Makefile:46]: Is a slash missing?

File: package-apisix.sh
Comment (EnableAsync) [inline package-apisix.sh:6]: Is a slash missing?

File: Makefile
Comment (imjoey) [reply Makefile]: > Is a slash missing?

@EnableAsync good catch and updated. Thanks.

File: package-apisix.sh
Comment (imjoey) [reply package-apisix.sh]: > Is a slash missing?

@EnableAsync updated. Thanks.

**Comment by imjoey** [comment]
After some digs, I find that `fpm` does not automatically add `--deb-dist` value to the name of deb file. Also, there seems no conventional way to show the `dist` value for deb file. So the first new commit reverts the deb filename at testing stage.

In addition, the second new commit sets the dist mandatory only for apisix, which could keep the CI jobs of apisix-dashboard and apisix-openresty successful.

@EnableAsync PTAL at your free time. Thanks.

**Comment by EnableAsync** [comment]
> After some digs, I find that `fpm` does not automatically add `--deb-dist` value to the name of deb file. Also, there seems no conventional way to show the `dist` value for deb file. So the first new commit reverts the deb filename at testing stage.
> 
> In addition, the second new commit sets the dist mandatory only for apisix, which could keep the CI jobs of apisix-dashboard and apisix-openresty successful.
> 
> @EnableAsync PTAL at your free time. Thanks.


You are right, fpm only adds dist tag to the rpm file name, maybe we can consider renaming it in our own way after packaging?

IMO, we can can rename it via `docker cp` in makefile. For example:
```bash
docker cp output:/output/apisix_2.2-0_amd64.deb $PWD/output/$(name)_$(version)-$(iteration)~$(dist)_$(arch).deb
```
What do you think of this idea? @imjoey @spacewander

File: Makefile
Comment (EnableAsync) [inline Makefile:34]: CI will fail when dist is empty.

**Comment by imjoey** [comment]
> You are right, fpm only adds dist tag to the rpm file name, maybe we can consider renaming it in our own way after packaging?
> 
> IMO, we can rename it via `docker cp` or add a new function for renaming in makefile. For example:
> 
> ```shell
> docker cp output:/output/apisix_2.2-0_amd64.deb $PWD/output/$(name)_$(version)-$(iteration)~$(dist)_$(arch).deb
> ```
> 
> What do you think of this idea? @imjoey @spacewander

@EnableAsync hey bro, I just find that we can have fine-grained control on the filename via the `-p` flag of `fpm` command. For example, if we use `fpm -s dir -t deb -n jenkins -v 2.305 --iteration 1 -p ./output/jenkins_2.305-2~focal1_amd64.deb jenkins.war`, then the deb file named `jenkins_2.305-2~focal1_amd64.deb` will be generated under `./output` directory. So the problem could be resolved via composing the filename with the variables, instead of depending on the `--deb-dist` or `--rpm-dist` flags. What do you think?

**Comment by EnableAsync** [comment]
> > You are right, fpm only adds dist tag to the rpm file name, maybe we can consider renaming it in our own way after packaging?
> > IMO, we can rename it via `docker cp` or add a new function for renaming in makefile. For example:
> > ```shell
> > docker cp output:/output/apisix_2.2-0_amd64.deb $PWD/output/$(name)_$(version)-$(iteration)~$(dist)_$(arch).deb
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What do you think of this idea? @imjoey @spacewander
> 
> @EnableAsync hey bro, I just find that we can have fine-grained control on the filename via the `-p` flag of `fpm` command. For example, if we use `fpm -s dir -t deb -n jenkins -v 2.305 --iteration 1 -p ./output/jenkins_2.305-2~focal1_amd64.deb jenkins.war`, then the deb file named `jenkins_2.305-2~focal1_amd64.deb` will be generated under `./output` directory. So the problem could be resolved via composing the filename with the variables, instead of depending on the `--deb-dist` or `--rpm-dist` flags. What do you think?

I think it's a great idea. 👍

File: Makefile
Comment (EnableAsync) [reply Makefile:34]: I think it is ok to support both apisix-dashboard and apisix-openresty.

**Comment by imjoey** [comment]
@EnableAsync 

Hi, another idea bumped into my mind this morning. Since users could specify the `image_base` and `image_tag` parameters while building and packaging. Why not determine the dist according to the OS distribution type of building container?

Furthermore, previous solution has a bug that if users run `make package type=rpm app=apisix version=2.2 checkout=2.2 dist=el8`, the `apisix-2.2-1.el8.rpm` is actually generated within the default `centos:7` container. So we should avoid users inputting the `dist` manually.

So I just refactor the code with a branch new force-push commit, doing the following changes:

- Add a new script `determine-dist.sh` under `./utils`, which can tell what the `dist` is by `image_base` and `image_tag` and write it into `/tmp/dist`.
- During the packaging phrase, the `/tmp/dist` is copied from building container to the packaging container, and then pass it as an argument to the `fpm` command.
- After `fpm` command, do the renaming if it's a deb artifact.

IMHO, this solution would keep the changes rather simple and intuitive. ping @EnableAsync for reviewing. Thanks.

> BTW, due to my poor network, I can hardly do the local testing related to this PR, so may I take use of the official CI for testing? Really sorry for this. Thanks.

File: Makefile
Comment (imjoey) [reply Makefile]: @EnableAsync yep, the latest commit has added all the three packages support. Please take a look at the new implementation https://github.com/api7/apisix-build-tools/pull/67#issuecomment-896536098. Really thanks for your help.

**Comment by EnableAsync** [comment]
> @EnableAsync
> 
> Hi, another idea bumped into my mind this morning. Since users could specify the `image_base` and `image_tag` parameters while building and packaging. Why not determine the dist according to the OS distribution type of building container?
> 
> Furthermore, previous solution has a bug that if users run `make package type=rpm app=apisix version=2.2 checkout=2.2 dist=el8`, the `apisix-2.2-1.el8.rpm` is actually generated within the default `centos:7` container. So we should avoid users inputting the `dist` manually.
> 
> So I just refactor the code with a branch new force-push commit, doing the following changes:
> 
>     * Add a new script `determine-dist.sh` under `./utils`, which can tell what the `dist` is by `image_base` and `image_tag` and write it into `/tmp/dist`.
> 
>     * During the packaging phase, the `/tmp/dist` is copied from building container to the packaging container, and then pass it as an argument to the `fpm` command.
> 
>     * After `fpm` command, do the renaming if it's a deb artifact.
> 
> 
> IMHO, this solution would keep the changes rather simple and intuitive. ping @EnableAsync for reviewing. Thanks.
> 
> > BTW, due to my poor network, I can hardly do the local testing related to this PR, so may I take use of the official CI for testing? Really sorry for this. Thanks.

Take it easy, it's not your fault.

In order not to wait for approval every time, you can create a new branch and modify the workflow file in your own repo to run the github action locally.
Like this: https://github.com/EnableAsync/apisix-build-tools/blob/imjoey/fpm_support_rpm_dist/.github/workflows/package-apisix-deb-focal1.yml#L5

Further, you can debug using tmate on your own branch which allows you to interact with Github Action.
Link: https://github.com/mxschmitt/action-tmate

File: Makefile
Comment (EnableAsync) [inline Makefile:31]: Is a slash missing?

Comment (imjoey) [reply Makefile]: 😢 sorry for this. Updated. Thanks.

**Comment by imjoey** [comment]
> Take it easy, it's not your fault.
> 
> In order not to wait for approval every time, you can create a new branch and modify the workflow file in your own repo to run the github action locally.
> Like this: https://github.com/EnableAsync/apisix-build-tools/blob/imjoey/fpm_support_rpm_dist/.github/workflows/package-apisix-deb-focal1.yml#L5
> 
> Further, you can debug using tmate on your own branch which allows you to interact with Github Action.
> Link: https://github.com/mxschmitt/action-tmate

@EnableAsync  I will try and very much appreciated your kind help. 😄

**Comment by EnableAsync** [comment]
Fails when I execute `make package type=deb app=apisix version=2.2 checkout=2.2`, and centos image is pulled by default. For deb, do we must add `image_base` and `image_tag` ?

**Comment by imjoey** [comment]
> Fails when I execute `make package type=deb app=apisix version=2.2 checkout=2.2`, and centos image is pulled by default. For deb, do we must add `image_base` and `image_tag` ?

@EnableAsync not really. It's a bug and we should provide defaults images for deb. I will make a change.

**Comment by imjoey** [comment]
> > Fails when I execute `make package type=deb app=apisix version=2.2 checkout=2.2`, and centos image is pulled by default. For deb, do we must add `image_base` and `image_tag` ?
> 
> @EnableAsync not really. It's a bug and we should provide defaults images for deb. I will make a change.

@EnableAsync  Hi, PR updated and the default image for deb is `ubuntu:focal`. I also add two CIs with el8 for rpm and default for deb to checks.  It's ready for review. Much appreicated. 😄

**Comment by EnableAsync** [comment]
LGTM.
@spacewander Please take a look.

File: test/apisix/Dockerfile.test.apisix.deb.focal1
Comment (spacewander) [inline test/apisix/Dockerfile.test.apisix.deb.focal1:11]: The package looks strange to me. Is there any info about the name convention?

File: .github/workflows/package-apisix-deb-focal1.yml
Comment (spacewander) [inline .github/workflows/package-apisix-deb-focal1.yml:1]: There are two workflows for ubuntu 20.04?

File: test/apisix/Dockerfile.test.apisix.deb.focal1
Comment (imjoey) [reply test/apisix/Dockerfile.test.apisix.deb.focal1:11]: > The package looks strange to me. Is there any info about the name convention?

@spacewander actually deb packages do not have a conventional name format for distribution section. So just in order to let users know which distribution is the package for, I refer to the openresty packages, see https://openresty.org/package/ubuntu/pool/main/o/openresty/, to define the apisix packages.

Looking forward to your insights. 😄

Comment (spacewander) [reply test/apisix/Dockerfile.test.apisix.deb.focal1:11]: Look like deb doesn't have a compulsory revision format: https://www.debian.org/doc/debian-policy/ch-controlfields.html#version

I think our revision should contain ubuntu or debian so that we can distinguish it.

File: .github/workflows/package-apisix-deb-focal1.yml
Comment (imjoey) [reply .github/workflows/package-apisix-deb-focal1.yml:1]: > There are two workflows for ubuntu 20.04?

yep, one is explicitly for ubuntu 20.04, the other is for testing the default image of `deb`.

Comment (imjoey) [reply .github/workflows/package-apisix-deb-focal1.yml:1]: I would propose turning on more workflows for more OS distributions in other PRs.

File: test/apisix/Dockerfile.test.apisix.deb.focal1
Comment (imjoey) [reply test/apisix/Dockerfile.test.apisix.deb.focal1:11]: @spacewander Agreed. We could use `package_1.2.3-0~debian11_amd64.deb` for the package on Debian 11. As currently Debian is not supported yet, so we can add this in other PRs.

Regarding the packages on Ubuntu, how about keeping using the name `package_1.2.3-0~focal1_amd64`? IMO our users usually know what `focal` means. 😄

What do you think?

Comment (spacewander) [reply test/apisix/Dockerfile.test.apisix.deb.focal1:11]: Since deb can be used in debian it is better to add distribution name in the version.

Comment (imjoey) [reply test/apisix/Dockerfile.test.apisix.deb.focal1:11]: > Since deb can be used in debian it is better to add distribution name in the version.

@spacewander OK, it's reasonable. How about `package_1.2.3-0~ubuntu20.04_amd64.deb`, which is the format that some packages are using. I could hardly see any packages with the name `package_1.2.3-0~ubuntufocal_amd64.deb`.

Looking forward to your feedback.

Comment (imjoey) [reply test/apisix/Dockerfile.test.apisix.deb.focal1]: @spacewander

Hi, I updated the PR and set deb name to `package_1.2.3-0~ubuntu20.04_amd64.deb` for Ubuntu 20.04, so `package_1.2.3-0~ubuntu21.04_amd64.deb` will be for Ubuntu 21.04. All CIs are passed within https://github.com/imjoey/apisix-build-tools/pull/2, please take a look at your free time.

**Comment by imjoey** [comment]
Also, ping @Yiyiyimu @membphis @nic-chen for reviewing. Thanks.

**Review by spacewander — APPROVED** [review]

**Pull request merged**
